src/main/java/org/apache/groovy/ast/tools/ExpressionUtils.java

```diff
@@ -18,20 +18,205 @@
+        handledTypes.add(DIVIDE);
+    }
+
+    public static ConstantExpression transformBinaryConstantExpression(BinaryExpression be, ClassNode targetType) {
+        if (isTypeOrArrayOfType(targetType, ClassHelper.STRING_TYPE, false)) {
+            if (be.getOperation().getType() == PLUS) {
+                Expression left = transformInlineConstants(be.getLeftExpression(), targetType);
+                Expression right = transformInlineConstants(be.getRightExpression(), targetType);
+                if (left instanceof ConstantExpression && right instanceof ConstantExpression) {
+                    ConstantExpression newExp = new ConstantExpression((String) ((ConstantExpression) left).getValue() +
+                            ((ConstantExpression) right).getValue());
+                    newExp.setSourcePosition(be);
+                    return newExp;
+                }
+            }
+        } else if (isTypeOrArrayOfType(targetType, ClassHelper.Integer_TYPE, false) || isTypeOrArrayOfType(targetType, ClassHelper.int_TYPE, false)) {
+            int type = be.getOperation().getType();
+            if (handledTypes.contains(type)) {
+                Expression left = transformInlineConstants(be.getLeftExpression(), targetType);
+                Expression right = transformInlineConstants(be.getRightExpression(), targetType);
+                if (left instanceof ConstantExpression && right instanceof ConstantExpression) {
+                    Integer newVal = null;
+                    switch(type) {
+                        case PLUS:
+                            newVal = (Integer) ((ConstantExpression) left).getValue() +
+                                    (Integer) ((ConstantExpression) right).getValue();
+                            break;
+                        case MINUS:
+                            newVal = (Integer) ((ConstantExpression) left).getValue() -
+                                    (Integer) ((ConstantExpression) right).getValue();
+                            break;
+                        case MULTIPLY:
+                            newVal = (Integer) ((ConstantExpression) left).getValue() *
+                                    (Integer) ((ConstantExpression) right).getValue();
+                            break;
+                        case DIVIDE:
+                            newVal = (Integer) ((ConstantExpression) left).getValue() /
+                                    (Integer) ((ConstantExpression) right).getValue();
+                            break;
+                    }
+                    if (newVal != null) {
+                        ConstantExpression newExp = new ConstantExpression(newVal, true);
+                        newExp.setSourcePosition(be);
+                        return newExp;
+                    }
+                }
+            }
+        } else if (isTypeOrArrayOfType(targetType, ClassHelper.Double_TYPE, false) || isTypeOrArrayOfType(targetType, ClassHelper.double_TYPE, false)) {
```

Comment:
looks like some template code, which is duplicated now. It's better to refactor it IMO

Reply:
Already refactored - thanks for the suggestion.